### ISSUE TITLE
[Feature Request] Hessian support for model weights (nn.Parameter)

### ISSUE THREAD

**Issue by ain-soph** [pytorch/pytorch#50138]
## 🚀 Feature
`torch.autograd.functional.hessian` provides a convenient way to calculate hessian for a function wrt an input. However, many times we don't want to calculate hessian wrt x (input), but hessian wrt θ (model parameters).  
Currently hessian API in this case have some confusion and not very convenient. There is a related post in discussion board:  
https://discuss.pytorch.org/t/using-autograd-functional-jacobian-hessian-with-respect-to-nn-module-parameters/103994/3  

--------------------------------------------------------------

## API Expectation
Currently, the following code returns a matrix filled with zero. (shape is correct `[4, 3, 4, 3]`)
```python3
import torch
import torch.nn as nn

_input=torch.randn(32,3)
layer = nn.Linear(3,4)
criterion=nn.CrossEntropyLoss()

weight = layer.weight

def func(weight):
    return criterion(layer(_input), torch.zeros(len(_input),dtype=torch.long))

torch.autograd.functional.hessian(func,weight)
```

## Current Workaround
Currently my solution is (which works as a workaround)
```python3
import torch
import torch.nn as nn

_input=torch.randn(32,3)
layer = nn.Linear(3,4)
criterion=nn.CrossEntropyLoss()

weight = layer.weight

def func(weight):
    del layer.weight
    layer.weight=weight
    return criterion(layer(_input), torch.zeros(len(_input),dtype=torch.long))

torch.autograd.functional.hessian(func,weight)
```
The drawback of this workaround is it's **Not Safe**: You have to delete the attribute `weight` in the `func` **everytime**, and can't be written just before `func` for only once. Otherwise, the hessian will all be 0.  

## API suggestion
According to #49171, `from functools import partial` already solve this problem. No need to do such modification.  
~~Even if the expected API works as above, it still seems not quite convenient since I need to define `func` a lot of times because my inputs are changing, but I can only define `func(weight)` rather than `func(input)` since hessian is wrt weight rather than input. This is caused by `func` only accepting one argument. What if the API changes like this:~~
```python3
def func(weight, _input, _label):
    return criterion(layer(_input), _label))

torch.autograd.functional.hessian(func, weight, _input, _label)
```
~~which means `func` might accept many argument of `*args` and `**kwargs`, which could be passed from `torch.autograd.functional.hessian(func, *args, **kwargs)`, but only `args[0]` will be taken as the variable to compute gradient.~~

-------------------------------------------  

Finally, I notice one thing that `torch.autograd.functional.hessian` doesn't require the `input` attribute `requires_grad` to be `True`, which is quite different from `torch.autograd.grad`. Is that a designed feature in purpose?

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

**Comment by albanD** [comment]
Hi,

Yes this design difference is done on purpose.
The plan we have to solve this issue is to build on top of https://github.com/pytorch/pytorch/issues/49171 so that the user can give the weights as proper inputs to the function and thus work well with the function autograd API.

**Comment by albanD** [comment]
Closing as a duplicate of https://github.com/pytorch/pytorch/issues/40480 as the title is more informative.
